plugins/autofarm.cpp

```diff
@@ -281,44 +268,33 @@ class AutoFarm {
 
         lastCounts.clear();
 
-        df::item_flags bad_flags;
-        bad_flags.whole = 0;
-
-#define F(x) bad_flags.bits.x = true;
-        F(dump); F(forbid); F(garbage_collect);
-        F(hostile); F(on_fire); F(rotten); F(trader);
-        F(in_building); F(construction); F(artifact);
-#undef F
+        const uint32_t bad_flags{
+ #define F(x) (df::item_flags::Mask::mask_##x)
+         F(dump) | F(forbid) | F(garbage_collect) |
+         F(hostile) | F(on_fire) | F(rotten) | F(trader) |
+         F(in_building) | F(construction) | F(artifact)
+ #undef F
+        };
 
         // have to scan both items[PLANT] and items[PLANT_GROWTH] because agricultural products can be either
 
-        for (auto ii : world->items.other[df::items_other_id::PLANT])
-        {
-            auto i = virtual_cast<df::item_plantst>(ii);
-            if (i &&
-                (i->flags.whole & bad_flags.whole) == 0 &&
-                plantable_plants.count(i->mat_index) > 0)
+        auto count = [&, this](df::item* i) {
+            auto mat = i->getMaterialIndex();
+            if ((i->flags.whole & bad_flags) == 0 &&
+                plantable_plants.count(mat) > 0)
             {
-                lastCounts[i->mat_index] += i->stack_size;
+                lastCounts[mat] += i->getStackSize();
             }
-        }
-
-        for (auto ii : world->items.other[df::items_other_id::PLANT_GROWTH])
-        {
-            auto i = virtual_cast<df::item_plant_growthst>(ii);
-            if (i &&
-                (i->flags.whole & bad_flags.whole) == 0 &&
-                plantable_plants.count(i->mat_index) > 0)
-            {
-                lastCounts[i->mat_index] += i->stack_size;
-            }
-        }
+        };
 
+        for (auto& i : world->items.other[df::items_other_id::PLANT])
```

Comment:
Elegant!

Reply:
it annoyed me that i had repeated code. i'm a big fan of DRY